Add examples for tutorial deployments

natalian98 opened the pull request:
### Summary
edit:
As discussed I added an `examples` directory that deploys the Identity Platform and its dependencies in 2 juju models.
The main module remains unchanged, the scope of this PR is to have a simple way for a user to start with the tutorial.

### Testing

To run the tutorial:

```
terraform -chdir=examples/tutorial init
terraform -chdir=examples/tutorial apply
```

This will create `core` model with dependencies and `iam` with the minimal setup for identity platform.

To deploy the extra charms:

Add `vars.tfvars` file:
```
model = "iam"
enable_admin_ui = true
enable_kratos_external_idp_integrator = true

idp_provider_config = {
  client_id   = "client-id"
  provider    = "provider-name"
  provider_id = "provider-id"
}

idp_provider_credentials = {
  client_secret = "some-secret"
}
```

Run the `extra` module:
```
terraform -chdir=examples/tutorial plan -var-file="./vars.tfvars"
terraform -chdir=examples/tutorial apply -var-file="./vars.tfvars"
```

This will deploy idp-integrator and admin-ui in the iam model.

nsklikas commented on README.md:
```diff
@@ -35,41 +35,37 @@ juju add-model core
-model = identity-platform
+postgresql_offer_url = "admin/core.pg-database"
+ingress_offer_url = "admin/core.ingress"
+send_ca_certificate_offer_url = "admin/core.send-ca-cert"
```
I thought the plan was to have an `examples` dir that would bring all these modules together without requiring extra steps from the user.

nsklikas commented on README.md:
````diff
@@ -105,6 +92,15 @@ Run `juju switch <juju model>` to switch to the target Juju model.
+If the extra charms were not deployed as part of the platform in the previous step, they can be added later with:
+```
+terraform -chdir=modules/extra init
+terraform -chdir=modules/extra apply -var-file="../../vars.tfvars" -var deploy_extra_charms=true
````
We shouldn't have to apply the same vars file to the `iam` and the `extra` modules.

nsklikas commented on modules/extra/variables.tf:
```diff
@@ -0,0 +1,94 @@
+  default     = "identity-platform"
+}
+
+variable "deploy_extra_charms" {
```
I don't think this needs to be part of the `extra` module.

nsklikas commented on modules/extra/variables.tf:
```diff
@@ -0,0 +1,94 @@
+  default     = "admin/core.send-ca-cert"
+}
+
+variable "kratos_name" {
```
This makes the assumption that the extra module will be deployed on the same model as kratos/hydra. We should at least document it somewhere (or make it work in the general case)

natalian98 replied:
added info in description

natalian98 replied:
https://github.com/canonical/iam-bundle-integration/blob/e6c2718d2057d72ba6c400158000065e6806fadb/modules/extra/variables.tf#L81

is this enough?

nsklikas commented on variables.tf:
```diff
@@ -4,6 +4,12 @@ variable "model" {
   default     = "identity-platform"
 }
 
+variable "deploy_extra_charms" {
+  type        = bool
+  default     = false
+  description = "Whether to deploy Identity Platform extensions: Admin UI, External IdP integrator, OpenFGA"
+}
```
I think this should be removed, I think the idea is that you can create a new tf plan and combine these 2 modules there.

natalian98 commented on modules/extra/applications.tf:
```diff
@@ -0,0 +1,40 @@
+module "kratos_external_idp_integrator" {
```
Docs will be updated, though I'm not aware of the issues you mentioned. Could you open a bug in kratos if you have more details?

natalian98 commented:
@nsklikas @wood-push-melon Models are now created by terraform. I added examples directory and moved the modules to it, so the root module remains unchanged and the PR is focused on simplifying the tutorial. Please have a look

wood-push-melon commented on examples/core/integrations.tf:
```diff
@@ -0,0 +1,40 @@
+  model            = juju_model.core.name
+  name             = "ingress"
+  application_name = juju_application.traefik.name
+  endpoints        = ["ingress"]
```
A minor thing

Just would like to double confirm with you:

Should it be `endpoint` or `endpoints`? The [doc](https://registry.terraform.io/providers/juju/juju/0.20.0/docs/resources/offer)'s example uses `endpoint` but the `Schema` section says `endpoints`.

wood-push-melon commented on examples/tutorial/outputs.tf:
```diff
@@ -0,0 +1,29 @@
+output "postgresql_offer_url" {
```
Just a nitpick (feel free to ignore):

It'd be nice to provide some descriptions.

wood-push-melon commented on examples/core/integrations.tf:
```diff
@@ -0,0 +1,40 @@
+  model = juju_model.core.name
+  application {
+    name     = juju_application.traefik.name
+    endpoint = "receive-ca-cert"
```
Should be `certificates`? Traefik needs to get a certificate signed by the `self-signed-certificates` charm, right?

wood-push-melon commented:
(feel free to ignore) Just a personal thought:

About how to organize the TF modules, I guess it'd be nicer to move `core` and `extra` under the `modules` since they are just "building blocks" to compose a working IAM platform. For example, `tutorial` only uses `core` since it just wants to quickly bootstrap a working IAM setup. We might have another example (maybe called `complete`) to use `core` and `extra` to bootstrap a more complete setup.

shipperizer commented on examples/tutorial/applications.tf:
```diff
@@ -0,0 +1,50 @@
+  }
+
+  config = var.kratos.config
+
+  depends_on = [juju_model.iam]
+}
+
+resource "juju_application" "hydra" {
+  model = juju_model.iam.name
+  name  = "hydra"
+  trust = var.hydra.trust
+  units = var.hydra.units
+
+  charm {
+    name    = "hydra"
+    channel = var.hydra.channel
+    base    = var.hydra.base
+  }
+
+  config = var.hydra.config
+
+  depends_on = [juju_model.iam]
+}
+
+resource "juju_application" "login_ui" {
+  model = juju_model.iam.name
+  name  = "login-ui"
+  trust = var.login_ui.trust
+  units = var.login_ui.units
+
+  charm {
+    name    = "identity-platform-login-ui-operator"
+    channel = var.login_ui.channel
+    base    = var.login_ui.base
+  }
+
+  config = var.login_ui.config
+
+  depends_on = [juju_model.iam, juju_application.hydra, juju_application.kratos]
+}
```
```suggestion
resource "juju_model" "this" {
  name = "${var.environment}-iam"
}


module "iam-bundle" {
  source = "../"
  model  = juju_model.this.name
  idp_provider_config = merge(
    var.idp_provider_config,
    {
      client_id : sensitive(data.vault_generic_secret.idp.data["client-id"]),
    },
  )
  idp_provider_credentials = {
    client_secret : sensitive(data.vault_generic_secret.idp.data["client-secret"])
  }

  postgresql_offer_url          = data.terraform_remote_state.core.outputs.database_offer
  ingress_offer_url             = data.terraform_remote_state.core.outputs.ingress_offer
  openfga_offer_url             = data.terraform_remote_state.core.outputs.openfga_offer
  send_ca_certificate_offer_url = data.terraform_remote_state.core.outputs.send_ca_certificate_offer

  hydra    = var.hydra
  kratos   = var.kratos
  login_ui = var.login_ui
  admin_ui = var.admin_ui
}


```

shipperizer replied:
we want to reuse the existing module on the root of the repo

natalian98 commented on modules/core/applications.tf:
```diff
@@ -0,0 +1,50 @@
+  }
+
+  config = var.certificates.config
+
+  depends_on = [juju_model.core]
+}
+
+resource "juju_application" "traefik" {
+  model = juju_model.core.name
+  name  = "traefik-public"
+  trust = var.traefik.trust
+  units = var.traefik.units
+
+  charm {
+    name    = "traefik-k8s"
+    channel = var.traefik.channel
+    base    = var.traefik.base
+  }
+
+  config = var.traefik.config
+
+  depends_on = [juju_model.core, juju_application.certificates]
+}
+
+resource "juju_application" "postgresql" {
+  model = juju_model.core.name
+  name  = "postgresql-k8s"
+  trust = var.postgresql.trust
+  units = var.postgresql.units
+
+  charm {
+    name    = "postgresql-k8s"
+    channel = var.postgresql.channel
+    base    = var.postgresql.base
+  }
+
+  config = var.postgresql.config
+
+  depends_on = [juju_model.core]
+}
```
wouldn't this cause problems if their `required_providers` or version differ from ours?

nsklikas commented on README.md:
This file needs to be updated, should it use the tutorial?

natalian98 replied:
You are right, sorry I missed that. Updated the readme.

natalian98 commented on examples/tutorial/main.tf:
```diff
@@ -0,0 +1,85 @@
+  postgresql_offer_url          = juju_offer.postgresql_offer.url
+  ingress_offer_url             = juju_offer.ingress_offer.url
+  send_ca_certificate_offer_url = juju_offer.send_ca_certificate_offer.url
+  openfga_offer_url             = var.enable_admin_ui && length(juju_offer.openfga_offer) > 0 ? juju_offer.openfga_offer[0].url : null
```
:+1:, removed `&& length(juju_offer.openfga_offer) > 0`
